RoddyCore/src/de/dkfz/roddy/execution/io/SSHExecutionService.groovy

```diff
@@ -353,83 +361,85 @@ class SSHExecutionService extends RemoteExecutionService {
         return result.toString("UTF-8")
     }
 
-    // This method actually overrides a base class. But if we keep the @Override, the Groovy (or Java) compiler constantly
-    // claims, that the method does not override it's base method.
-    // That is, why we keep it in but only as a comment.
-    //    @Override
-    synchronized ExecutionResult _execute(String command,
-                                          boolean waitFor = true,
-                                          boolean ignoreError = false,
-                                          OutputStream outputStream = null) {
-        SSHPoolConnectionSet connectionSet = waitForService()
-        SSHClient sshClient = connectionSet.client
+    private static final ExecutorService executorService = Executors.newCachedThreadPool()
 
-        if (waitFor) {
+    @Override
+    ExecutionResult _execute(String command,
+                             boolean waitFor = true,
+                             Duration timeout = Duration.ZERO,
+                             OutputStream outputStream = null) {
+        _execute(command, waitFor, timeout, outputStream, executorService)
+    }
+
+    /**
+     * An com.hierynomus:sshj:0.23.0 and JSCH Agent 0.0.9 (-> net.schmizz:sshj:0.10.0) compatibility
+     * problem prevents the usage of Command.join(long, TimeUnit) from working. The best solution is
+     * probably to get rid of the whole SSHJ/JSCH stack and instead use mina-sshd, which is not
+     * completely outdated and actively maintained (and can use the SSH-agent and execute with timeouts.
+     *
+     * The solution taken now uses the Linux command `timeout`.
+     *
+     * @param command
+     * @param waitFor
+     * @param timeout
+     * @param outputStream
+     * @param executorService
+     * @return
+     */
+    ExecutionResult _execute(String command,
+                             boolean waitFor = true,
+                             Duration timeout = Duration.ZERO,
+                             OutputStream outputStream = null,
+                             ExecutorService executorService) {
+
+        CompletableFuture<Tuple3<Integer,List<String>, List<String>>> processF = CompletableFuture.supplyAsync({
+            SSHPoolConnectionSet connectionSet = waitForService()
+            SSHClient sshClient = connectionSet.client
+
+            // Append a newgrp (Philip: better "newgrp -"!) to each command, so that all
+            // command context in the proper group context.
```

Comment:
This comment doesn't make sense to me, where is `newgrp` added (and shouldn't it be prepended)?

Otherwise your changes look good

Reply:
I think the comment is not from me and you're right in that it is misplaced and unclear. I will just remove it. I remember the discussion, though. The aim was to run commands such that files are already created with the right group. We never followed up on this.